triage/berghelroach/berghelroach_test.go

```diff
@@ -0,0 +1,342 @@
+}
+
+/*
+verifyResult verifies a single edit distance result.
+If the expected distance is within limit, result must b
+be correct; otherwise, result must be over limit.
+
+s1: one string compared
+
+s2: other string compared
+
+expectedResult: correct distance from s1 to s2
+
+k: limit applied to computation
+
+d: distance computed
+*/
+func verifyResult(t *testing.T, s1 string, s2 string, expectedResult int, k int, d int) {
+	// Check if the strings are long. If so, they're not worth printing.
+	var s1Printed string
+	var s2Printed string
+	if utils.Max(len(s1), len(s2)) > 50 {
+		s1Printed = "string1"
+		s2Printed = "string2"
+	} else {
+		s1Printed = s1
+		s2Printed = s2
+	}
+
+	if k >= expectedResult {
+		if expectedResult != d {
+			t.Errorf("Distance from %s to %s should be %d (within limit=%d) but was %d", s1Printed, s2Printed, expectedResult, k, d)
+		}
+	} else {
+		if !(d > k) {
+			t.Errorf("Distance from %s to %s should be %d (exceeding limit=%d) but was %d", s1Printed, s2Printed, expectedResult, k, d)
+		}
+	}
+}
+
+// genericVerification exercises an edit distance engine across a wide range of limit values
+func genericVerification(t *testing.T, ed berghelRoach, s1 string, s2 string, expectedResult int) {
+	if len(s1) < 500 {
+		// For small strings, try every limit
+		maxDiff := utils.Max(len(s1), len(s2)) + 2
+		for k := 0; k < maxDiff; k++ {
+			verifyResult(t, s1, s2, expectedResult, k, ed.getDistance(s2, k))
+		}
+	} else {
+		// For big strings, try a sampling of limits:
+		//   0 to 3,
+		//   another 4 on either side of the expected result
+		//   s2 length
+		for k := 0; k < 4; k++ {
+			verifyResult(t, s1, s2, expectedResult, k, ed.getDistance(s2, k))
+		}
+		for k := utils.Max(4, expectedResult-4); k < expectedResult+4; k++ {
+			verifyResult(t, s1, s2, expectedResult, k, ed.getDistance(s2, k))
+		}
+		verifyResult(t, s1, s2, expectedResult, len(s2),
+			ed.getDistance(s2, len(s2)))
+	}
+
+	// Always try near MAX_VALUE
+	if ed.getDistance(s2, int(^uint(0)>>1)-1) != expectedResult { // int(^uint(0) >> 1) is the largest possible int
+		t.Errorf("getDistance failed with maximum uint value minus 1, got %d", expectedResult)
+	}
+	if ed.getDistance(s2, int(^uint(0)>>1)) != expectedResult { // int(^uint(0) >> 1) is the largest possible int
+		t.Errorf("getDistance failed with maximum uint value, got %d", expectedResult)
+	}
+}
+
+// specificAlgorithmVerify tests a specific engine on a pair of strings
+func specificAlgorithmVerify(t *testing.T, ed berghelRoach, s1 string, s2 string, expectedResult int) {
+	genericVerification(t, ed, s1, s2, expectedResult)
+
+	// Try again with the same instance
+	genericVerification(t, ed, s1, s2, expectedResult)
+}
+
+// genericLevenshteinVerify tests the default Levenshtein engine on a pair of strings
+func genericLevenshteinVerify(t *testing.T, s1 string, s2 string, expectedResult int) {
+	specificAlgorithmVerify(t, getInstance(s1), s1, s2, expectedResult)
+}
+
```

Comment:
nit: for followup, I'm not sure this extra level of indirection is worth it, I'd probably collapse the two together

Reply:
Addressed in #18881